Fix Issue and Add Scroll View

ambarwootric opened the pull request:
Changes:
1. For Issue: https://github.com/Wootric/WootricSDK-Android/issues/50
- Using Fragment (FragmentManager, FragmentActivity & DialogFragment) from Support Library v4.
- Trello: https://trello.com/c/y6MsJeXx/2755-use-fragement-from-support-library-dialogfragment-fragmentmanager

2. For Issue: https://github.com/Wootric/WootricSDK-Android/issues/49
- Add TAG `SURVEY_DIALOG_TAG` to Fragment Transaction somewhat https://github.com/Wootric/WootricSDK-Android/blob/master/androidsdk/src/main/java/com/wootric/androidsdk/SurveyManager.java#L187
- Trello: https://trello.com/c/fplSIQQ7/1112-add-tag-to-fragment-transaction

3. Fix `Survey view getting hidden from bottom if Survey question get's long` by adding Scroll View
- Please refer to changes of `https://github.com/Wootric/WootricSDK-Android/compare/aa-fix-issue-and-scroll-view?expand=1#diff-31822127d59d919a5fbe722716e1493f`
- The issue was not only with German Translation, it's more because if lengthy NPS question is found. 
- Trello: https://trello.com/c/FPX4S5Zd/2780-android-buttons-and-feedback-field-hidden-when-language-is-long-ie-german-translation

Testing:
- Go to https://github.com/Wootric/WootricSDK-Android/blob/master/app/src/main/java/com/wootric/androidsdk_demo/MainActivity.java and update your credential or refer to http://docs.wootric.com/android/#installing-android-sdk to setup your app for testing purpose and start a survey.
- For seeing Scroll View update surve language to German like `wootric.setLanguageCode("DE");` as it converts to have long NPS question. Or, you can update NPS question by setting up Product Name like ` wootric.setProductName("this is very very longgggg product name that I am using for testing the view of message bar of our product in android so let's have fun");`

Images of Scroll View showing Fix:
1. For German Language:
On Start:
<img width="184" alt="android-scroll-view-gr-fix-one" src="https://user-images.githubusercontent.com/28694156/35549259-cba09a76-05ac-11e8-9110-7a589b06fe41.PNG">
On Scroll:
<img width="184" alt="android-scroll-view-gr-fix-two" src="https://user-images.githubusercontent.com/28694156/35549282-ecc37e3a-05ac-11e8-8d69-ab2eb146263c.PNG">

2. For Long Product Name:
On Start:
<img width="184" alt="android-scroll-view-whole-one" src="https://user-images.githubusercontent.com/28694156/35549304-0a7445cc-05ad-11e8-96da-e2d8f2aa3e94.PNG">
On Scroll:
<img width="184" alt="android-scroll-view-whole-two" src="https://user-images.githubusercontent.com/28694156/35549321-1736279e-05ad-11e8-9569-1c803cfd1e8b.PNG">

Note: I also have updated build version `27.0.3` along with min SDK version `16` and target SDK version `26`

diegoserranoa commented on androidsdk/build.gradle:
```diff
@@ -1,12 +1,12 @@
+    buildToolsVersion '27.0.3'
 
     defaultConfig {
-        minSdkVersion 14
```
This is required to `support` libraries, right? Might wanna mention this in the mobile channel in Slack. Changing the minimum version means we're going to lose support for older devices...

ambarwootric replied:
@diegoserranoa I gave a thought on that but found that we already are mentioning to support for `minSdkVersion 16`. If we see here https://github.com/Wootric/WootricSDK-Android/blob/master/README.md under `Requirements` section.

diegoserranoa replied:
Ah you're right. That probably was an error on the docs side. But still, if we had `minSdkVersion 14` it means that we still supported devices with version 14 & 15. I would still mention this in the channel...

diegoserranoa commented on androidsdk/build.gradle:
```diff
@@ -42,13 +42,14 @@ def isReleaseBuild() {
-    testCompile 'org.assertj:assertj-core:1.7.0'
-    testCompile 'org.mockito:mockito-core:1.9.5'
-    compile 'com.android.support.constraint:constraint-layout:1.0.2'
+    implementation fileTree(dir: 'libs', include: ['*.jar'])
```
🙌  gradle 3.0

diegoserranoa commented on androidsdk/src/main/java/com/wootric/androidsdk/SurveyManager.java:
```diff
@@ -23,7 +23,8 @@
 package com.wootric.androidsdk;
 
 import android.app.Activity;
```
Is this still needed?

diegoserranoa commented on androidsdk/src/main/java/com/wootric/androidsdk/SurveyManager.java:
```diff
@@ -184,7 +185,7 @@ private void showSurveyFragment() {
             if(isTablet) {
                 fragmentManager.beginTransaction()
-                        .add(android.R.id.content, surveyFragment)
+                        .add(android.R.id.content, surveyFragment, SURVEY_DIALOG_TAG)
```
👍

diegoserranoa commented on androidsdk/src/main/java/com/wootric/androidsdk/views/SurveyFragment.java:
```diff
@@ -24,7 +24,7 @@
 
 import android.app.Activity;
 import android.app.Dialog;
```
Is this still needed?

diegoserranoa commented on androidsdk/src/test/java/com/wootric/androidsdk/SurveyManagerTest.java:
```diff
@@ -1,6 +1,7 @@
 package com.wootric.androidsdk;
 
 import android.app.Activity;
```
Is this still needed?

diegoserranoa commented on build.gradle:
```diff
@@ -16,5 +17,6 @@ buildscript {
 allprojects {
     repositories {
         jcenter()
+        google()
```
What is this for?

ambarwootric replied:
Diego this is needed for support library we are using from google maven repository i.e. `google()`. The latest versions and future support for android related libraries now will be used from google maven repository in gradle 4+ otherwise it will give me build error. More on here https://developer.android.com/studio/build/dependencies.html#google-maven

diegoserranoa replied:
Thanks for the explanation and the link!

diegoserranoa commented on app/src/main/java/com/wootric/androidsdk_demo/MainActivity.java:
```diff
@@ -2,11 +2,12 @@
 import com.wootric.androidsdk.Wootric;
 
-public class MainActivity extends Activity {
+public class MainActivity extends FragmentActivity {
```
Don't really need to modify this. But if you can do it if you want to. Take it or leave it...

DanPitrowiski commented:
I approve this design :+1:

---

I wanted to provide some footnotes here though, just for context on my approval.

(1) From Android's design guidelines point of view, to make this design work, we would have our SEND and DISMISS buttons fixed to the bottom of the component.
https://material.io/guidelines/components/dialogs.html#dialogs-specs

![components-dialogs-confirmation2](https://user-images.githubusercontent.com/2568493/35642192-2da914c6-0677-11e8-8bbc-05bc2c44e598.png)

For Wootric though, this doesn't quite work. We want to keep our "Powered by Wootric" at the bottom of our surveys. If we wanted to make SEND and DISMISS buttons fixed to the bottom, that would mean "Powered by Wootric" would need to float awkwardly in the middle of our survey. Or would need some funny custom UI that moved it around as the survey height changed. 

Ambar's solution here works with those constraints.

(2) If we had a way to track response rates for Android Surveys where the question gets long (and this UI is used), I would say we 100% should. Just to make sure we are maintaining our high response rates. It probably won't be an issue, but I wanted to note that.

diegoserranoa commented:
@DanPitrowiski 
(1) Wouldn't it be better to keep the `Powered by Wootric` and SEND and DISMISS buttons both fixed at the bottom? Just an idea :) but if this design also works then I say we go with this and don't put more time in it since this fixes the issue and it's just for small devices.

(2) The problem with this is where would we track that?

Another idea: wouldn't it look better if we just make the text scrollable instead of the whole view? That way we can keep the actions (i.e. the scores, buttons...) fixed.

ambarwootric commented:
@diegoserranoa
1. Yes, that will take a whole design to new other level and will have to reconsider with new designs so better is now.
2. I have tried so but fixing the size for QUESTION container will lead to having a design issue with feedback layout. As that container is dynamic and works well with current constraint.

ambarwootric commented:
Thanks for your questions and suggestions @diegoserranoa. I have tried to resolve some things and in the meantime when you will be testing, I will also push changes with README and CHANGELOG etc.
1. I used `Optimize Imports` feature of Android Studio to resolve unused imports and placement.
2. Hopefully, we are going with current scroll design. I DM @DanPitrowiski and provided him demo video and also we all went through some discussion. Got approved by Dan.
3. Talked with @prabhatjha and we are good to go with updating `minSdkVersion` to `16`. Although, I found that `Fragment` Support Library still can work with `minSdkVerstion 11` but let's not take chances and upgrade things up.

DanPitrowiski commented:
> (1) Wouldn't it be better to keep the Powered by Wootric and SEND and DISMISS buttons both fixed at the bottom? Just an idea :) but if this design also works then I say we go with this and don't put more time in it since this fixes the issue and it's just for small devices.

Yeah @diegoserranoa,  this would be cool. I haven't seen it done with an Android pop-over component before in my research. Especially for it to work dynamically based upon the size. 

So I am worried we might break something. Or end up down a rabbit hole where we put it a lot more effort and it doesn't fix the problem.

Do you think it's more easily possible?



> (2) The problem with this is where would we track that?

Yeah we'd have to add some tracking data to our surveys. I'm not exactly sure yet. But it's not worth the effort yet.

> Another idea: wouldn't it look better if we just make the text scrollable instead of the whole view?

Yeah this sounds like a good idea too. It' another design I haven't really seen used before so I wasn't sure if it would break it :upside_down_face:. But @ambarwootric seems to think it would make it harder for the survey to be dynamic.

ambarwootric commented:
@diegoserranoa thank you so much for your explanation and review. 
Thanks for letting me know about such BG issue and suggestion. I finally was able to replicate it with API level 16 and screen size 720x1280 and is now fixed (Image below):

<img width="165" alt="gray-bg-with-lower-api-fix" src="https://user-images.githubusercontent.com/28694156/35691207-f64a2982-079f-11e8-8634-c18b827f9d98.PNG">

Also, updated PR with changes required to show versioning and deployment.
